File: src/ai_novel_workstation/config/paths.py
"""应用路径与用户配置目录解析。

打包为 exe 后，`__file__` 指向临时解包目录（PyInstaller 单文件模式），
不能作为配置文件存放位置。所有用户配置统一放在系统用户配置目录
（Windows: %APPDATA% ；其他平台: ~/.config ），与程序文件隔离，升级不丢。
"""

from __future__ import annotations

import os
from pathlib import Path

from loguru import logger

# 应用名（用于用户配置目录名）
_APP_DIR_NAME = "AI-Novel-Workstation"

# 程序根目录（源码运行时 = 仓库根；打包后 = 解包临时目录，仅用于定位只读资源）
_PROJECT_ROOT = Path(__file__).resolve().parents[3]


def get_user_config_dir() -> Path:
    """返回用户配置根目录（存放 .env 与运行时数据）。

    可用环境变量 AI_NOVEL_CONFIG_DIR 覆盖（测试/便携模式用）。
    """
    override = os.environ.get("AI_NOVEL_CONFIG_DIR")
    if override:
        return Path(override).expanduser().resolve()

    if os.name == "nt":
        base = os.environ.get("APPDATA") or str(Path.home() / "AppData" / "Roaming")
    else:
        base = str(Path.home() / ".config")
    return Path(base) / _APP_DIR_NAME


def get_env_file_path() -> Path:
    """返回用户配置目录下的 .env 路径（确保目录存在）。"""
    d = get_user_config_dir()
    d.mkdir(parents=True, exist_ok=True)
    return d / ".env"


def get_legacy_env_path() -> Path:
    """旧版本的程序根 .env（用于首次运行迁移）。"""
    return _PROJECT_ROOT / ".env"


def get_project_root() -> Path:
    """程序根目录（只读资源定位用，非配置存放位置）。"""
    return _PROJECT_ROOT


def migrate_env_if_needed() -> Path:
    """首次运行迁移：若新位置无 .env 而旧位置（程序根）有，则复制。

    Returns:
        实际使用的 .env 路径
    """
    target = get_env_file_path()
    if target.exists():
        return target

    legacy = get_legacy_env_path()
    if legacy.exists():
        try:
            target.parent.mkdir(parents=True, exist_ok=True)
            target.write_bytes(legacy.read_bytes())
            logger.info(f"已迁移配置 .env → {target}")
        except Exception as e:
            logger.warning(f"迁移 .env 失败: {e}")
    return target

File: backend/src/config/paths.ts
/**
 * 应用路径与用户配置目录解析（TS 版，迁移自 config/paths.py）。
 * 所有用户配置统一放在系统用户配置目录（Windows: %APPDATA%；其他平台: ~/.config），
 * 与程序文件隔离，升级不丢。
 */
import * as fs from "node:fs";
import * as os from "node:os";
import { resolve } from "node:path";
import { fileURLToPath } from "node:url";

export const APP_DIR_NAME = "AI-Novel-Workstation";

/** 源码/编译入口定位：backend/src/config（dev）或 backend/dist/config（prod） */
export const BACKEND_ROOT = resolve(fileURLToPath(import.meta.url), "../../..");
/** 仓库根（数据目录 data/ 所在位置） */
export const PROJECT_ROOT = resolve(BACKEND_ROOT, "..");

export function getUserConfigDir(): string {
  const override = process.env.AI_NOVEL_CONFIG_DIR;
  if (override) return resolve(override);
  if (process.platform === "win32" && process.env.APPDATA) {
    return resolve(process.env.APPDATA, APP_DIR_NAME);
  }
  return resolve(os.homedir(), ".config", APP_DIR_NAME);
}

export function getEnvFilePath(): string {
  const dir = getUserConfigDir();
  fs.mkdirSync(dir, { recursive: true });
  return resolve(dir, ".env");
}

export function getLegacyEnvPath(): string {
  return resolve(PROJECT_ROOT, ".env");
}

/** 运行期数据目录（app-state / interactions / agent-roles / projects）。桌面版可用 AI_NOVEL_DATA_DIR 覆盖到用户可写目录。 */
export function getDataDir(): string {
  const override = process.env.AI_NOVEL_DATA_DIR;
  if (override) return resolve(override);
  return resolve(PROJECT_ROOT, "data");
}

/** 本地 embedding 模型目录（离线权重）。桌面版可用 AI_NOVEL_EMBEDDING_DIR 覆盖到内置资源目录。 */
export function getEmbeddingDir(): string {
  const override = process.env.AI_NOVEL_EMBEDDING_DIR;
  if (override) return resolve(override);
  return resolve(getUserConfigDir(), "models");
}

export function migrateEnvIfNeeded(): string {
  const target = getEnvFilePath();
  if (fs.existsSync(target)) return target;
  const legacy = getLegacyEnvPath();
  if (fs.existsSync(legacy)) {
    try {
      fs.mkdirSync(resolve(target, ".."), { recursive: true });
      fs.copyFileSync(legacy, target);
      console.log(`[paths] 已迁移配置 .env -> ${target}`);
    } catch (err) {
      console.warn(`[paths] 迁移 .env 失败: ${err}`);
    }
  }
  return target;
}
